src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,309 @@
+
+  // use the chain rule to compute the first-order change in the q-function output
+  auto dq = chain_rule(dq_darg, darg);
+
+  // integrate dq against test space shape functions / gradients
+  // to get the (change in) element residual contributions
+  dr_elem += Postprocess<test_element>(dq, xi, J_q) * dx;
+
+}
+  
+  
+template <Geometry g, typename test, typename trial, int Q,
+          typename derivatives_type, typename du_type, typename dr_type>
+__global__ void gradient_cuda_element(const du_type du, dr_type dr, derivatives_type* derivatives_ptr,
+                     const mfem::DeviceTensor<4, const double> J, int num_elements)
+{
+  using test_element               = finite_element<g, test>;
+  using trial_element              = finite_element<g, trial>;
+  using element_residual_type      = typename trial_element::residual_type;
+  static constexpr auto rule       = GaussQuadratureRule<g, Q>();
+
+
+  // for each element in the domain
+  //  for (int e = 0; e < num_elements; e++) {
+  //int e = blockIdx.x * blockDim.x + threadIdx.x;
+  //if (e < num_elements) {
+  const int grid_stride = blockDim.x * gridDim.x;
+#pragma unroll
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x; e < num_elements; e += grid_stride) {
+    // get the (change in) values for this particular element
+    tensor du_elem = detail::Load<trial_element>(du, e);
+
+    // this is where we will accumulate the (change in) element residual tensor
+    element_residual_type dr_elem{};
+
+    // for each quadrature point in the element
+    for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+
+      gradient_quadrature<g, test, trial, Q, derivatives_type>(e, q, du_elem, dr_elem, derivatives_ptr, J, num_elements);
+    }
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(dr, dr_elem, e);
+  }
+}
+
+template <Geometry g, typename test, typename trial, int Q,
+          typename derivatives_type, typename du_type, typename dr_type>
+__global__ void gradient_cuda_quadrature(const du_type du, dr_type dr, derivatives_type* derivatives_ptr,
+                     const mfem::DeviceTensor<4, const double> J, int num_elements)
+{
+  using test_element               = finite_element<g, test>;
+  using trial_element              = finite_element<g, trial>;
+  using element_residual_type      = typename trial_element::residual_type;
+  static constexpr auto rule       = GaussQuadratureRule<g, Q>();
+
+
+  // for each element in the domain
+  //  for (int e = 0; e < num_elements; e++) {
+  // int qe = blockIdx.x * blockDim.x + threadIdx.x;
+  // if (qe < num_elements * rule.size()) {
+
+  const int grid_stride = blockDim.x * gridDim.x;
+#pragma unroll
+  for (int qe = blockIdx.x * blockDim.x + threadIdx.x; qe < num_elements * rule.size(); qe += grid_stride) {
```

Comment:
it might be clearer to move the start/end values outside the loop
```suggestion
  auto thread_id = blockIdx.x * blockDim.x + threadIdx.x;
  auto num_quadrature_points = num_elements * rule.size();
  for (int qe = thread_id; qe < num_quadrature_points; qe += grid_stride) {
```